### PR TITLE
#826: Create a basic, accessible, reusable map component (no boundaries)

### PR DESCRIPTION
**Pull request by joegasewicz**
+ #849: Map controls (zoom, full screen, reset map state)
### What
Creates basic reusable map component with support for Ordnance Survey tiles in the DP design system repo

Describe what you have changed and why.
Added a `map.js` file which setups a Maxbox.js map.
### How to review

Describe the steps required to test the changes.
  ```     
    <div 
       id="map"
       data-style="mapbox://styles/mapbox/cjaudgl840gn32rnrepcb9b9g?access_code=<key>"
       data-center="[51.42133794405771, -0.14646169096877365]"
       data-zoom="3"
   ></div> 
 ```

### Who can review
Any frontend developers familiar with this library

**#849: Map controls (zoom, full screen, reset map state):**

Describe who worked on the changes, so that other people can review.
myself

- Implement the zoom, reset zoom and full screen controls in the base map component in the dp-design-system library ✅
- Be able to move around the map ✅
- When moving around the map I should not be able to scroll further than the boundary of the UK ✅
- User should be able to zoom in ✅
  - When zooming in more detail should be shown on the map) ✅
  - Zoom in should be a smooth transition (i.e. not snap to zoom) ✅
- Zoom out ✅
  - When zooming out less detail should be shown on the map, but major reference points should still be shown for positional reference ) ✅
  - When zooming out I should not be able to zoom out further than a view with the UK filling the map ✅
  - Zoom out should be a smooth transition (i.e. not snap to zoom) ✅
- Be able to make the map full screen and then return to smaller version ✅

- On hover over the controls should display a badge with helpful text ✅
- Zoom should work with touch (pinch to zoom), mouse wheel, keyboard and the buttons ✅
- Pan should work with touch, click and drag and keyboard navigation ✅ _(touch left & right increments magnification only)_
- Disable pitch controls (i.e. should always be straight down) ✅

- Add a reset to default state button and keyboard shortcut ✅
  - If we are in the starting position (pan and zoom) then the "Reset Zoom" control will be disabled (greyed out). As soon as the map state is changed the button should be enabled  ✅
  - When clicking the reset the map should change as a smooth transition (i.e. fly to area and zoom level rather than snap to) ✅

- Focus state:
  - When tabbing through page and we tab to the map the map should have a focus state (yellow border around it, see designs) ✅
  - Continuing to tab will bring you to the map controls which will then have the yellow border focus state (see designs)  ✅

-screen reader

- Help text for keyboard navigation (TBC, speak to Trevor) to be displayed with map ⚠️

### CODE REVIEW

File: src/js/dp/map.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    // ID
+    const mapIDName = "map";
+    const mapElem = document.getElementById(mapIDName);
+    if (!mapElem) {
+        throw new Error("To use the OSD Map Component you must declare an id of 'map' on the containing HTML element");
+    }
+    // Style
+    const mapStyle = _getMapStyle(mapElem);
+    // Zoom
+    let mapZoom = _getMapZoom(mapElem);
```
Comment (joegasewicz) [inline src/js/dp/map.js:51]: This should have been a const

File: src/js/dp/map.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ *       id="map"
+ *       data-style="mapbox://styles/mapbox/cjaudgl840gn32rnrepcb9b9g?access_code=<key>"
+ *       data-center="[51.42133794405771, -0.14646169096877365]"
+ *       data-zoom="3"
+ *   ></div> 
+ * ```
+ */
+function initMap() {
+    window = window || {};
+    console.log("Loading map...")
```
Comment (joegasewicz) [inline src/js/dp/map.js:80]: needs removing

**Comment by gilesburdett** [comment]
Would it be preferrable to move the images into the `cdn.ons.gov.uk-assets` repository?

https://github.com/ONSdigital/cdn.ons.gov.uk-assets/tree/master/assets/images

File: src/js/dp/map.js [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    return mapStyle
+};
+/** @internal */
+const _getMapCenter = (mapEl) => {
+    let mapCenter = mapEl.dataset.center;
+    if (!mapCenter) {
+        throw new Error("You must declare a 'data-center' attribute");
+    }
+    mapCenter = mapCenter.replace(/\[|]/g, "");
+    if (mapCenter.length < 2) throw new Error(centerErrorMsg);
```
Comment (gilesburdett) [inline src/js/dp/map.js:37]:
```suggestion
    if (mapCenter.length < 2) {
        throw new Error(centerErrorMsg);
    }
```

File: src/js/dp/map.js [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    mapCenter = mapCenter.split(",");
+    mapCenter = mapCenter.map(center => {
```
Comment (gilesburdett) [inline src/js/dp/map.js:39]: Counterintuitively the radix does not default to 10:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt#description

```suggestion
    mapCenter = mapCenter.map(center => parseInt(center, 10));
```

File: src/js/dp/map.js [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    return mapCenter;
+};
+/** @internal */
+const _getMapZoom = (mapEl) => {
+    let mapZoom = parseInt(mapEl.dataset.zoom)
+    if (!mapZoom) {
+        mapZoom = 3;
+        console.warn("OSD Map Component: defaulting the zoom value to 3");
+    } else {
+        mapZoom =  parseInt(mapEl.dataset.zoom)
```
Comment (gilesburdett) [inline src/js/dp/map.js:51]:
```suggestion
        mapZoom =  parseInt(mapEl.dataset.zoom, 10);
```

File: src/js/dp/map.js [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    if (mapCenter.length < 2) throw new Error(centerErrorMsg);
+    mapCenter = mapCenter.split(",");
+    mapCenter = mapCenter.map(center => {
+        return parseInt(center);
+    });
+    return mapCenter;
+};
+/** @internal */
+const _getMapZoom = (mapEl) => {
+    let mapZoom = parseInt(mapEl.dataset.zoom)
```
Comment (gilesburdett) [inline src/js/dp/map.js:46]:
```suggestion
    let mapZoom = parseInt(mapEl.dataset.zoom, 10);
```

File: src/js/dp/map.js [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    if (!mapZoom) {
+        mapZoom = 3;
+        console.warn("OSD Map Component: defaulting the zoom value to 3");
+    } else {
+        mapZoom =  parseInt(mapEl.dataset.zoom)
+    }
+    return mapZoom;
+};
+/** @internal */
+const _getToken = (mapEl) => {
```
Comment (gilesburdett) [inline src/js/dp/map.js:56]: const _getToken = (mapEl) => mapEl.dataset.token;

File: src/js/dp/map.js [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+    constructor(center, zoom, mapState) {
+        this.mapState = mapState;
+        this.center = center;
+        this.zoom = zoom;
+    }
+
+    set controlElements(val) {
+        if (!this._controlElements.length) {
+            this._controlElements = val.map((elem, i) => {
```
Comment (gilesburdett) [inline src/js/dp/map.js:92]: Unused index

```suggestion
            this._controlElements = val.map((elem) => {
```

File: src/js/dp/map.js [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        e.preventDefault();
+        const canvas = this.map.getCanvas();
+        if (e.which === Keys.R) {
+            this._onReset(e);
+            return;
+        }
+        if (e.which === Keys.ESCAPE) {
+            canvas.blur();
+        }
+
```
Comment (gilesburdett) [inline src/js/dp/map.js:139]:
```suggestion
switch (e.which) {
    case Keys.R:
        this._onReset(e);
        break;
    case Keys.Escape:
        canvas.blur(); 
        break;
}
```

or `if` ... `else if` without the `return` in the first block if `switch` isn't the house style

**Review by gilesburdett — CHANGES_REQUESTED** [review]

**Comment by gilesburdett** [comment]
Will this work with 2+ maps on a page, or is it expected that that would never occur?

**Comment by mr-nick17** [comment]
My main concern is that this implementation will load all the map js and css for pages that instantiate the `dp-design-system` regardless of whether there is a map on the page. What's the difference in payload?

**Comment by joegasewicz** [comment]
> My main concern is that this implementation will load all the map js and css for pages that instantiate the `dp-design-system` regardless of whether there is a map on the page. What's the difference in payload?

Solution for this is coming in another pr, will update with a link asap, thank you

**Pull request closed without merging**